Cython/Compiler/Nodes.py

```diff
@@ -4962,14 +4962,22 @@ def generate_execution_code(self, code):
             code.putln("else {")
         else:
             code.putln("else if (")
-            code.putln("#if CYTHON_USE_TYPE_SLOTS || CYTHON_COMPILING_IN_PYPY")
-            code.putln(f"unlikely(Py_TYPE({self_arg})->tp_dictoffset != 0)")
+            code.putln("#if !(CYTHON_USE_TYPE_SLOTS || CYTHON_COMPILING_IN_PYPY)")
+            # If CYTHON_USE_TYPE_SPECS then all extension types are heap-types so the check below automatically
+            # passes and thus takes the slow route.
+            # Therefore we do a less thorough check - if they type hasn't changed then clearly it hasn't
+            # been overridden, and if the type isn't GC then it also won't have been overridden.
+            if self.py_func.entry.scope.lookup_here("__dict__"):
+                code.putln("(1)")
```

Comment:
Isn't this something general that could go next to the `is_module_scope` above, i.e. use no C code condition at all?

Reply:
Moved.... I was originally trying not to change the non-limited-api path at all, but this should be fairly minor.